transform_constantfold.ml

```diff
@@ -43,6 +43,9 @@ let const_prop ({formals; instrs} : analysis_input) : instructions option =
     | Assign (y, e) ->
       assert (x <> y);
       Assign (y, replace e)
+    | Array_assign (y, i, e) ->
+      assert (x <> y);
```

Comment:
I'm a bit suspicious. (I guess this is partly because constant folding is being changed in a new PR, and we're also considering changing the semantics of arrays.)

I believe the assert will be triggered in the following program:
```
const x = [1]
x[0] <- 0
```

The function `find_candidates` will see the constant definition `const x = [1]` and add it as a "candidate," and then try to do the constant propagation here, where the assert will trigger.